optimum/exporters/executorch/integrations.py

```diff
@@ -26,13 +26,440 @@
+                )
+            else:
+                # Manually set the attention implementation to custom_sdpa_ring_kv_cache
+                # This handles both regular sdpa and one for sliding window/local attention
+                exportable_module.model.model.config._attn_implementation = (
+                    "custom_sdpa"
+                )
+
+    def export(
+        self,
+    ) -> Dict[str, ExportedProgram]:
+
+        exportable_module = TorchExportableModuleForImageTextLM(
+            self.model,
+            max_batch_size=1,
+            max_cache_len=self.metadata.get("get_max_seq_len"),
+        )
+        self._register_attention_mask_for_4_53(exportable_module)
+
+        if self.use_custom_kv_cache:
+            from optimum.executorch.attentions.custom_kv_cache import (
+                replace_with_et_custom_kv_cache,
+            )
+
+            replace_with_et_custom_kv_cache(
+                exportable_module.model,
+                self.model.config.text_config,
+                self.model.generation_config,
+                self.model.dtype,
+            )
+
+        with torch.no_grad():
+            inputs_embeds, cache_position, dynamic_shapes, strict = (
+                self._prepare_decoder_only_export_inputs()
+            )
+            logging.info(
+                f"Exporting decoder using inputs_embeds({inputs_embeds.shape}), cache_position({cache_position.shape})={cache_position}, dynamic_shapes={dynamic_shapes}, strict={strict}"
+            )
+            exported_program = exportable_module.export(
+                inputs_embeds, cache_position, dynamic_shapes, strict
+            )
+            # Apply RemoveTransposes pass to remove
+            # any back-to-back transpose ops that are not needed
+            # e.g. output of update_cache is transposed and
+            # input to custom_sdpa is transposed.
+            from executorch.extension.llm.export.export_passes import (
+                RemoveRedundantTransposes,
+            )
+
+            mutated_gm = RemoveRedundantTransposes()(exported_program.module())[0]
+            exported_program = torch.export.export(
+                mutated_gm,
+                args=(),
+                kwargs={
+                    "cache_position": cache_position,
+                    "inputs_embeds": inputs_embeds,
+                },
+                dynamic_shapes=dynamic_shapes,
+                strict=strict,
+            )
+
+            # Export token embeddings
+            input_ids, dynamic_shapes, strict = (
+                self._prepare_text_embedding_export_inputs()
+            )
+            logging.info(
+                f"Exporting token embeddings using input_ids({input_ids.shape}), dynamic_shapes={dynamic_shapes}, strict={strict}"
+            )
+
+            token_embeddings_exported_program = torch.export.export(
+                exportable_module.model.model.language_model.get_input_embeddings(),
```

Comment:
I dont follow this. we already export exported_module.model. So I would have expected that `get_input_embedding` is traced as part of that? I guess thats not the case when `input_embeds != None`

Reply:
I guess that’s inevitable and I hope transformers folks can give some guarantees lol. Like model.vision_model and model.get_image_features()